java/sage/nio/BufferedFileChannel.java

```diff
@@ -0,0 +1,967 @@
+    {
+      // Don't allow more than what was requested to be written.
+      if (writeBuffer.limit() > count)
+        writeBuffer.limit((int)count);
+
+      bytesWritten = src.read(writeBuffer);
+
+      if (bytesWritten == -1)
+        break;
+
+      count -= bytesWritten;
+
+      if (!writeBuffer.hasRemaining())
+        flushWriteBuffer();
+    }
+    while (count > 0);
+    writeBuffer.limit(writeBuffer.capacity());
+
+    return position() - startPosition;
+  }
+
+  @Override
+  public int read(ByteBuffer dst, long position) throws IOException
+  {
+    // Pending writes need to be flushed or they might not show up when reading.
+    if (writePending)
+      flushWriteBuffer();
+    else if (readBuffer.limit() != 0)
+    {
+      long readPosition = position - (realFilePosition - readBuffer.limit());
+
+      // The buffer is in the way, use this to our advantage.
+      if (readPosition >= -dst.remaining() && readPosition < readBuffer.limit())
+      {
+        long startPosition = position;
+        long readBytes;
+
+        // The read starts before the buffer.
+        int dstLimit = dst.limit();
+
+        if (readPosition < 0)
+        {
+          try
+          {
+            dst.limit(dst.position () + (int)-readPosition);
+
+            while (true)
+            {
+              readBytes = fileChannel.read(dst, position);
+
+              if (readBytes == 0)
+              {
+                return (int) (position - startPosition);
+              }
+
+              position += readBytes;
+
+              if (position == realFilePosition)
+              {
+                break;
+              }
+            }
+          }
+          finally
+          {
+            dst.limit(dstLimit);
+          }
+        }
+
+        // The read starts somewhere in the buffer and possibly goes outside of it.
+        int readBufferPosition = readBuffer.position();
+        int readBufferLimit = readBuffer.limit();
+        int readLength = Math.min(readBuffer.remaining(), dst.remaining());
+
+        try
+        {
+          readBuffer.limit(readBufferPosition + readLength);
+          do
+          {
+            dst.put(readBuffer);
+          }
+          while (readBuffer.hasRemaining());
+          position += readLength;
+        }
+        finally
+        {
+          readBuffer.limit(readBufferLimit).position(readBufferPosition);
+        }
+
+        if (!dst.hasRemaining())
+          return (int)(position - startPosition);
+
+        return fileChannel.read(dst, position) + (int)(position - startPosition);
+      }
+    }
+
+    return fileChannel.read(dst, position);
+  }
+
+  @Override
+  public int write(ByteBuffer src, long position) throws IOException
+  {
+    if (readonly)
+      throw new NonWritableChannelException();
+
+    if (readBuffer.limit() != 0)
+      clearReadBuffer();
+    else if (writePending)
+    {
+      if (writeBuffer.position() < writeLength)
+        writeLength = writeBuffer.position();
```

Comment:
I'm not sure on this one...it seems by shortening writeLength here to match writeBuffer.position() that you are invalidating anything in the writeBuffer that's past the position but before the length. Is the comparator inverted here?